content/blog/2023/2023-06-25-archunit.md

```diff
@@ -0,0 +1,598 @@
+there's no dependency on any class in the persistence package. Let's see what happens if we introduce
+such a dependency:
+
+```Java
+@GET()
+@Path("/employees")
+public EmployeeResponse getEmployee() {
+  EmployeeDao dao = new EmployeeDao();
+  Employee employee = dao.findEmployee();
+  return new EmployeeResponse(
+    employee.id(),
+    employee.name(),
+    employee.active()
+  );
+}
+```
+
+With this code, we access the persistence layer directly in the controller class. As a result,
+the test will fail with an assertion error, informing us about the access violation:
+
+```Bash
+java.lang.AssertionError: Architecture Violation [Priority: MEDIUM] - Rule 'no classes should access classes that reside in a package 'io.reflectoring.archunit.persistence'' was violated (2 times):
+Method <io.reflectoring.archunit.api.EmployeeController.getEmployee()> calls constructor <io.reflectoring.archunit.persistence.EmployeeDao.<init>()> in (EmployeeController.java:15)
+Method <io.reflectoring.archunit.api.EmployeeController.getEmployee()> calls method <io.reflectoring.archunit.persistence.EmployeeDao.findEmployee()> in (EmployeeController.java:16)
+```
+
+The example shows how easy it is to use ArchUnit in a Java project. Before we look at more examples, let's discuss why
+architecture violations are typically introduced in projects over time. 
+
+## Reasons for Architecture Erosion Over Time
+There are many reasons why developers start to deviate from the initial design choices, coding best practices, or testing practices.
+One of the most common reasons is probably time pressure. As this is rather straightforward, we we'll look at some other reasons
+in more detail.
+
+### Architecture Awareness
+At the start of a software project, we usually take certain design choices and organize the code in methods, classes, 
+packages, modules, and layers. Each of these has its specific purpose and a clear boundary. 
+The data access layer for example should have the sole responsibility to retrieve persisted data.
+It should not provide an API endpoint or map data to an external format like JSON or XML. 
+
+We also make choices on certain implementation details like the inheritance (for example, every DAO class should implement
+an interface), or how to handle date and time in the code.
+
+Let's look at a simple example. Instead of:
+
+```java
+LocalDateTime localDate = LocalDateTime.now();
+```
+
+we might want to use:
+
+```java
+LocalDateTime localDate = LocalDateTime.now(clock);
+```
+
+When we decide to use the latter way of instantiating our object, we'll probably remember the reason for a while. However, after some
+time we might forget. Also, other developers who join the project might unintentionally deviate from the original choice. 
+
+With ArchUnit, we can add a test that will fail if the static factory method ```now``` is used without the parameter: 
+
+```java
+@Test
+public void alwayInjectClockToFaciliateTesting() { 
+  // ...
```

Comment:
I would expect the entire test in this place.

Reply:
Added!